### PR TITLE
Restore original share prototype

### PR DESCRIPTION
**Pull request by trevor-coleman**
This copies over the changes from Ryan's original prototype. This includes @lindboe's commits from https://github.com/infinitered/ExpensifyApp/pull/70. (They have been merged, so the diff should be clean)

### Remove temp hook
* Remove `useReactNativeShareMenuData` 
* Remove reference in `App.js`

### Create context to hold shared data and add to root
* create `ShareContext` and native version
* Add `Share.Provider` to the `ComposeProviders` array in `App.js`

### Create stack navigator for share flow
* Create`ShareModalStackNavigator`
* Added to `RightModalNavigator`
* Added to `ModalStackNavigator`
* Update `linkingConfig` -- `config.screens.[NAVIGATORS.RIGHT_MODAL_NAVIGATOR].screens.Share`
* Add `ROUTES.SHARE` and `ROUTES.SHARE_MESSAGE`

### Create SharePage and ShareMessagePage components
* Create Share Page 
* Create ShareMessagePage

### Update TabSelector to show options on Share Page
* Add option to `TabSelector` component
* Add `CONST.TAB.SHARE` and `CONST.TAB.SHARE_TAB_ID -- to hold the option for `TabSelector`

### Make `NewChatPage` navigate to ShareMessagePage when appropriate
* add a `navigateToAndOpenShare` action in `libas/actions/Report`
* check if share data exists when creating a new chat:
    *  if it does divert to `ShareMessagePage`
    * if not continue
* add dummy translations: `en.newChatPage.shareToExpensify` and `es.newChatPage.shareToExpensify`

### CODE REVIEW

**Comment by trevor-coleman** [comment]
This should be ready to go. At the very least it's working and a big enough chunk that any fixes or updates are probably better off being handled separately.

**Comment by lindboe** [comment]
It looks like something is going wrong trying to share an ogg file from the Files app on iOS (which is a test case that was working previously). Looks like it's probably crashign pretty early on. For reference, here's the file I tested with: https://upload.wikimedia.org/wikipedia/commons/e/e9/C_harmonic.ogg

Recording of unopenable Expensify

This doesn't look like the view delegate crash issue (which would block interaction, happen about every-other attempt, and didn't have animation like this). Also, if I go back and share an image from Photos again, that does open properly.

**Comment by lindboe** [comment]
Two issues with Android that I noticed that I don't think were there before:

1. Wrong file data displayed. I'm not 100% sure this wasn't an issue before. Attachment preview isn't working when it should, although it does work when the message is sent. In addition, the wrong file name is shown (looks like probably a temp file name? But we should be able to display the original). The wrong file name issue persists when the file is sent if it's a non-previewable file (I used C_harmonic.ogg in the below screenshot):

Screenshot 2023-11-18 at 3 45 16 PM

2. Sharing a file while the app is NOT in the background is not working. We never navigate to the share screen. I think whether we should fix this in this PR depends on the cause: (a) if it's that we aren't reading the share data correctly, we should fix that now. (b) if it's because routes aren't initialized yet and we couldn't navigate, we can tackle that later, especially since we want to compare in the deep link ticket.

**Comment by lindboe** [comment]
I'm seeing some warnings when starting up the app now, I think we should clear those out before merging this where we can. (I'm more concerned with "CAnnot read property `screen` of undefined" and "Looks like you're passing an inline function for "component" prop of the screen `scan`" than I am the a11y role warning).

File: patches/react-native-share-menu+6.0.0.patch
Comment (lindboe) [inline patches/react-native-share-menu+6.0.0.patch:61]: Let's remove this typing, the other keys don't use it

File: src/CONST.ts
Comment (lindboe) [inline src/CONST.ts:2709]: Should this come in as part of this PR?

File: src/pages/SharePage.js
Comment (lindboe) [inline src/pages/SharePage.js:58]: I noticed that Android is defaulting to the "Scan" tab, and iOS is defaulting to "Share". Not urgent to fix, but any idea why that might be happening?

**Review by lindboe — CHANGES_REQUESTED** [review]
I left a few notes on behavior that I don't recall happening on the old prototype. I think if we can look into those briefly it might help highlight if anything meaningful is different; otherwise we can turn them into tickets.

File: patches/react-native-share-menu+6.0.0.patch
Comment (trevor-coleman) [reply patches/react-native-share-menu+6.0.0.patch:61]:
```suggestion
+public let HOST_APP_GROUP_INFO_PLIST_KEY = "ReactNativeShareMenuAppGroup"
```

File: src/CONST.ts
Comment (trevor-coleman) [reply src/CONST.ts:2709]: The change to text needs to be there to prevent a crash, but not the new PRESENTATION key. The thing is this change *will* be breaking for other components until my ACCESSIBILITY_ROLE PR gets merged.
```suggestion
        TEXT: 'text',
```

Comment (lindboe) [reply src/CONST.ts:2709]: I'm leaning towards we should remove these. Your PR looks like it's dealing with a lot of merge conflicts, and I'd rather deal with type errors here than a bunch of merge conflicts later :) Thoughts?

File: patches/react-native-share-menu+6.0.0.patch
Comment (lindboe) [inline patches/react-native-share-menu+6.0.0.patch:104]: It looks like this patch is missing a few lines from the original:

1. More calls to `detachViewDelegate`, which Ryan added to avoid some crashes: https://github.com/infinitered/ExpensifyApp/pull/63/files#diff-1aa046d161a1081bd541198ac073c301a3773ce2f3591f5f88a5a6fc85c02e36R150. Not all of these might be necessary now.
2. file type handling: https://github.com/infinitered/ExpensifyApp/pull/63/files#diff-1aa046d161a1081bd541198ac073c301a3773ce2f3591f5f88a5a6fc85c02e36R168

@trevor-coleman Let me know if you have bandwidth to update this, if not, I'll do it tomorrow. Either way it might be good/less confusing to break this up into multiple patches since patch-package can do that now, which is exciting.

File: .github/CODEOWNERS
Comment (trevor-coleman) [inline .github/CODEOWNERS]: I'm guessing all the various markdown edits are an artifact of something upstream?

File: src/pages/SharePage.js
Comment (trevor-coleman) [inline src/pages/SharePage.js:1]: Was wondering where this was coming from. Nice.

**Review by trevor-coleman — COMMENTED** [review]
Changes look good! Apparently I can't approve my own PR, but if I could I would.

File: src/CONST.ts
Comment (trevor-coleman) [reply src/CONST.ts:2717]: If we remove this the app will crash.

File: src/pages/SharePage.js
Comment (lindboe) [reply src/pages/SharePage.js:1]: Very weird, this is not showing me anything:
Screenshot 2023-12-05 at 3 30 05 PM

**Pull request merged**
